File: authandshare/src/main/java/com/fox/authandshare/share/sina/SinaShare.java
package com.fox.authandshare.share.sina;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.Bundle;
import android.text.TextUtils;

import com.fox.authandshare.platform.PlatForm;
import com.fox.authandshare.share.IShare;
import com.fox.authandshare.share.ShareListener;
import com.fox.authandshare.share.ShareParamsHelper;
import com.fox.authandshare.share.ShareType;
import com.sina.weibo.sdk.WbSdk;
import com.sina.weibo.sdk.api.ImageObject;
import com.sina.weibo.sdk.api.TextObject;
import com.sina.weibo.sdk.api.VideoSourceObject;
import com.sina.weibo.sdk.api.WeiboMultiMessage;
import com.sina.weibo.sdk.auth.AuthInfo;
import com.sina.weibo.sdk.share.WbShareCallback;
import com.sina.weibo.sdk.share.WbShareHandler;

import java.io.File;

/**
 * Created by fox.hu on 2018/9/3.
 */

public class SinaShare implements IShare, WbShareCallback {
    private static final String TAG = SinaShare.class.getSimpleName();
    private WbShareHandler shareHandler;
    private ShareListener listener;
    private Context context;

    public SinaShare(Activity activity) {
        context = activity.getApplicationContext();
        WbSdk.install(activity,
                new AuthInfo(activity, PlatForm.SINA.getAppId(), PlatForm.SINA.getAppKey(),
                        PlatForm.SINA.getScope()));
        shareHandler = new WbShareHandler(activity);
        shareHandler.registerApp();
    }

    @Override
    public boolean isInstall(Context context) {
        return WbSdk.isWbInstall(context);
    }

    @Override
    public ShareParamsHelper getParamsHelper() {
        return new SinaShareHelper();
    }

    @Override
    public void shareTo(ShareType type, Activity activity, Bundle bundle, ShareListener listener) {
        this.listener = listener;
        WeiboMultiMessage weiboMessage = new WeiboMultiMessage();
        SinaShareBean bean = bundle.getParcelable(SinaShareHelper.SINA_BUNDLE);
        switch (type) {
            case TEXT:
                weiboMessage.textObject = getTextObj(bean);
                break;
            case IMAGE:
                weiboMessage.imageObject = getImageObj(bean);
                break;
            case VIDEO:
                weiboMessage.videoSourceObject = getVideoObject(bean);
                break;
            case APP:
                //todo 此处为了测试 之后删掉
                weiboMessage.textObject = getTextObj(bean);
                weiboMessage.imageObject = getImageObj(bean);
                weiboMessage.videoSourceObject = getVideoObject(bean);
                break;
            case MULTIPLE:
                weiboMessage.textObject = getTextObj(bean);
                weiboMessage.imageObject = getImageObj(bean);
                weiboMessage.videoSourceObject = getVideoObject(bean);
                break;
            default:
                break;
        }
        shareHandler.shareMessage(weiboMessage, false);
    }

    private VideoSourceObject getVideoObject(SinaShareBean bean) {
        String filePath = bean.getAudioPath();
        if (TextUtils.isEmpty(filePath)) {
            return null;
        } else {
            VideoSourceObject videoSourceObject = new VideoSourceObject();
            videoSourceObject.videoPath = Uri.fromFile(new File(filePath));
            return videoSourceObject;
        }
    }

    private ImageObject getImageObj(SinaShareBean bean) {
        int imgUrl = bean.getResId();
        if (imgUrl == 0) {
            return null;
        } else {
            ImageObject imageObject = new ImageObject();
            Bitmap bitmap = BitmapFactory.decodeResource(context.getResources(), imgUrl);
            imageObject.setImageObject(bitmap);
            return imageObject;
        }
    }

    private TextObject getTextObj(SinaShareBean bean) {
        TextObject textObject = new TextObject();
        textObject.text = bean.getText();
        textObject.title = bean.getTitle();
        textObject.actionUrl = bean.getActionUrl();
        return textObject;
    }

    @Override
    public void onNewIntent(Intent intent) {
        shareHandler.doResultIntent(intent, this);
    }

    @Override
    public void onActivityResultData(int requestCode, int resultCode, Intent data) {
    }

    @Override
    public void onWbShareSuccess() {
        listener.onComplete("share success!");
    }

    @Override
    public void onWbShareCancel() {
        listener.onCancel();
    }

    @Override
    public void onWbShareFail() {
        listener.onError("share failed");
    }
}

File: authandshare/src/main/java/com/fox/authandshare/share/sina/SinaShare.kt
package com.fox.authandshare.share.sina

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.graphics.BitmapFactory
import android.net.Uri
import android.os.Bundle
import android.text.TextUtils
import com.fox.authandshare.platform.PlatForm
import com.fox.authandshare.share.IShare
import com.fox.authandshare.share.ShareListener
import com.fox.authandshare.share.ShareParamsHelper
import com.fox.authandshare.share.ShareType
import com.sina.weibo.sdk.WbSdk
import com.sina.weibo.sdk.api.ImageObject
import com.sina.weibo.sdk.api.TextObject
import com.sina.weibo.sdk.api.VideoSourceObject
import com.sina.weibo.sdk.api.WeiboMultiMessage
import com.sina.weibo.sdk.auth.AuthInfo
import com.sina.weibo.sdk.share.WbShareCallback
import com.sina.weibo.sdk.share.WbShareHandler
import java.io.File

/**
 * Created by fox.hu on 2018/9/3.
 */

class SinaShare(activity: Activity) : IShare, WbShareCallback {
    private val shareHandler: WbShareHandler
    private var listener: ShareListener? = null
    private val context: Context

    override val paramsHelper: ShareParamsHelper<*>
        get() = SinaShareHelper()

    init {
        context = activity.applicationContext
        WbSdk.install(
            activity,
            AuthInfo(
                activity, PlatForm.SINA.appId, PlatForm.SINA.appKey,
                PlatForm.SINA.scope
            )
        )
        shareHandler = WbShareHandler(activity)
        shareHandler.registerApp()
    }

    override fun isInstall(context: Context): Boolean {
        return WbSdk.isWbInstall(context)
    }

    override fun shareTo(
        type: ShareType,
        activity: Activity,
        bundle: Bundle,
        listener: ShareListener
    ) {
        this.listener = listener
        val weiboMessage = WeiboMultiMessage()
        val bean = bundle.getParcelable<SinaShareBean>(SinaShareHelper.SINA_BUNDLE)
        when (type) {
            ShareType.TEXT -> weiboMessage.textObject = getTextObj(bean!!)
            ShareType.IMAGE -> weiboMessage.imageObject = getImageObj(bean!!)
            ShareType.VIDEO -> weiboMessage.videoSourceObject = getVideoObject(bean!!)
            ShareType.APP -> {
                //todo 此处为了测试 之后删掉
                weiboMessage.textObject = getTextObj(bean!!)
                weiboMessage.imageObject = getImageObj(bean)
                weiboMessage.videoSourceObject = getVideoObject(bean)
            }
            ShareType.MULTIPLE -> {
                weiboMessage.textObject = getTextObj(bean!!)
                weiboMessage.imageObject = getImageObj(bean)
                weiboMessage.videoSourceObject = getVideoObject(bean)
            }
            else -> {
            }
        }
        shareHandler.shareMessage(weiboMessage, false)
    }

    private fun getVideoObject(bean: SinaShareBean): VideoSourceObject? {
        val filePath = bean.audioPath
        if (TextUtils.isEmpty(filePath)) {
            return null
        } else {
            val videoSourceObject = VideoSourceObject()
            videoSourceObject.videoPath = Uri.fromFile(File(filePath!!))
            return videoSourceObject
        }
    }

    private fun getImageObj(bean: SinaShareBean): ImageObject? {
        val imgUrl = bean.resId
        if (imgUrl == 0) {
            return null
        } else {
            val imageObject = ImageObject()
            val bitmap = BitmapFactory.decodeResource(context.resources, imgUrl)
            imageObject.setImageObject(bitmap)
            return imageObject
        }
    }

    private fun getTextObj(bean: SinaShareBean): TextObject {
        val textObject = TextObject()
        textObject.text = bean.text
        textObject.title = bean.title
        textObject.actionUrl = bean.actionUrl
        return textObject
    }

    override fun onNewIntent(intent: Intent) {
        shareHandler.doResultIntent(intent, this)
    }

    override fun onActivityResultData(requestCode: Int, resultCode: Int, data: Intent) {}

    override fun onWbShareSuccess() {
        listener!!.onComplete("share success!")
    }

    override fun onWbShareCancel() {
        listener!!.onCancel()
    }

    override fun onWbShareFail() {
        listener!!.onError("share failed")
    }

    companion object {
        private val TAG = SinaShare::class.java.simpleName
    }
}
